src/ray/common/memory_monitor.cc

```diff
@@ -0,0 +1,205 @@
+        [this] {
+          bool is_usage_above_threshold = IsUsageAboveThreshold();
+          monitor_callback_(is_usage_above_threshold);
+        },
+        monitor_interval_ms,
+        "MemoryMonitor.CheckIsMemoryUsageAboveThreshold");
+    RAY_LOG(INFO) << "MemoryMonitor initialized";
+#else
+    RAY_LOG(WARNING) << "Not running MemoryMonitor. It is currently supported "
+                     << "only on Linux.";
+#endif
+  } else {
+    RAY_LOG(INFO) << "MemoryMonitor disabled. Specify "
+                  << "`memory_monitor_interval_ms` > 0 to enable the monitor.";
+  }
+}
+
+bool MemoryMonitor::IsUsageAboveThreshold() {
+  auto [used_memory_bytes, total_memory_bytes] = GetMemoryBytes();
+  if (total_memory_bytes == kNull || used_memory_bytes == kNull) {
+    RAY_LOG_EVERY_MS(WARNING, kLogIntervalMs)
+        << "Unable to capture node memory. Monitor will not be able "
+        << "to detect memory usage above threshold.";
+    return false;
+  }
+  auto usage_fraction = static_cast<float>(used_memory_bytes) / total_memory_bytes;
+  bool is_usage_above_threshold = usage_fraction > usage_threshold_;
+  if (is_usage_above_threshold) {
+    RAY_LOG_EVERY_MS(INFO, kLogIntervalMs)
+        << "Node memory usage above threshold, used: " << used_memory_bytes
+        << ", total: " << total_memory_bytes << ", usage fraction: " << usage_fraction
+        << ", threshold: " << usage_threshold_;
+  }
+  return is_usage_above_threshold;
+}
+
+std::tuple<int64_t, int64_t> MemoryMonitor::GetMemoryBytes() {
+  RAY_CHECK(__linux__) << "Memory monitor currently supports only linux";
+  auto [system_used_bytes, system_total_bytes] = GetLinuxMemoryBytes();
+  auto [cgroup_used_bytes, cgroup_total_bytes] = GetCGroupMemoryBytes();
+  /// cgroup memory limit can be higher than system memory limit when it is
+  /// not used. We take its value only when it is less than or equal to system memory
+  /// limit. TODO(clarng): find a better way to detect cgroup memory limit is used.
+  system_total_bytes = NullableMin(system_total_bytes, cgroup_total_bytes);
+  /// This assumes cgroup total bytes will look different than system (meminfo)
+  if (system_total_bytes == cgroup_total_bytes) {
+    system_used_bytes = cgroup_used_bytes;
+  }
+  return std::tuple(system_used_bytes, system_total_bytes);
+}
+
+std::tuple<int64_t, int64_t> MemoryMonitor::GetCGroupMemoryBytes() {
+  int64_t total_bytes = kNull;
+  if (boost::filesystem::exists(kCgroupsV2MemoryMaxPath)) {
+    std::ifstream mem_file(kCgroupsV2MemoryMaxPath, std::ios::in | std::ios::binary);
+    mem_file >> total_bytes;
+  } else if (boost::filesystem::exists(kCgroupsV1MemoryMaxPath)) {
+    std::ifstream mem_file(kCgroupsV1MemoryMaxPath, std::ios::in | std::ios::binary);
+    mem_file >> total_bytes;
+  }
+
+  int64_t used_bytes = kNull;
+  if (boost::filesystem::exists(kCgroupsV2MemoryUsagePath)) {
+    std::ifstream mem_file(kCgroupsV2MemoryUsagePath, std::ios::in | std::ios::binary);
```

Comment:
This [blog post](https://medium.com/@eng.mohamed.m.saeed/memory-working-set-vs-memory-rss-in-kubernetes-which-one-you-should-monitor-8ef77bf0acee) claims either high RSS or high WSS can get container OOM killed in k8s. IIUC the memory usage here is WSS. We can see how well the current scheme works and if we need to read other types of usages.